README.rst
**************************************
Make Your Own Neural Network with Ruby
**************************************

https://github.com/artob/myonn.rb

If.rb #5
--------

This project was first presented at the
`If.rb #5 <https://www.facebook.com/events/290820645150764/>`__
meetup in Ivano-Frankivsk, Ukraine on March 6, 2019.

The slide deck is available at:

https://speakerdeck.com/arto/make-your-own-neural-network-with-ruby-at-if-dot-rb-number-5


CHANGES.rst
*********
Changelog
*********

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog
<https://keepachangelog.com/en/1.0.0/>`__.

2019.03.06
==========

The initial version first presented at
`If.rb #5 <https://www.facebook.com/events/290820645150764/>`__
meetup in Ivano-Frankivsk, Ukraine.


mnist_test.rb
#!/usr/bin/env ruby -Ilib
# This is free and unencumbered software released into the public domain.

require 'myonn'
require 'mnist'

model = MNIST::Model.load('data/mnist_nn.yaml')

score = model.validate('data/mnist_test.csv')
puts '%.1f%%' % (score * 100.0)


lib/myonn.rb
# This is free and unencumbered software released into the public domain.

require 'numo/linalg/linalg'
Numo::Linalg::Loader.load_openblas '/opt/homebrew/Cellar/openblas/0.3.5/lib'
#puts Numo::Linalg::Loader.libs

require 'numo/narray'
require 'yaml'

module MYONN
  class NeuralNetwork
    attr_reader :input_nodes
    attr_reader :input_weights
    attr_reader :hidden_nodes
    attr_reader :output_weights
    attr_reader :output_nodes
    attr_reader :learning_rate
    attr_reader :activation_function

    def expit(x)
      (1 + Numo::NMath.exp(-x)) ** -1
    end

    def self.load(filepath)
      data = YAML.load_file(filepath)
      %i(input_weights output_weights).each do |k|
        v = data[k]
        data[k] = Numo::DFloat.cast(v['data']).reshape(*v['shape'])
      end
      self.new(data)
    end

    ##
    # Initializes the network.
    def initialize(input_nodes:, input_weights: nil, hidden_nodes:, output_nodes:, output_weights: nil, learning_rate: 0.3, activation_function: nil)
      @input_nodes         = input_nodes
      @input_weights       = input_weights  || Numo::DFloat.new(hidden_nodes, input_nodes).rand_norm(0.0, input_nodes ** -0.5)
      @hidden_nodes        = hidden_nodes
      @output_weights      = output_weights  || Numo::DFloat.new(output_nodes, hidden_nodes).rand_norm(0.0, hidden_nodes ** -0.5)
      @output_nodes        = output_nodes
      @learning_rate       = learning_rate.to_f
      @activation_function = method(activation_function || :expit)
    end

    ##
    # Returns a hash with this network's configuration.
    def to_hash
      {
        input_nodes: @input_nodes,
        hidden_nodes: @hidden_nodes,
        output_nodes: @output_nodes,
        learning_rate: @learning_rate,
        activation_function: @activation_function.name,
        input_weights: @input_weights,
        output_weights: @output_weights,
      }
    end

    ##
    # Returns a YAML string with this network's configuration.
    def to_yaml
      data = self.to_hash.inject({}) do |hash, (k, v)|
        hash[k] = case v
          when Numo::DFloat
            count = v.shape.reduce(1, :*)
            {
              'count' => count,
              'shape' => v.shape,
              'data'  => v.reshape(count).to_a,
            }
          else v
        end
        hash
      end
      YAML.dump(data)
    end

    def save(filepath)
      File.open(filepath, 'w') do |file|
        file.write(self.to_yaml)
      end
    end

    ##
    # Trains the network.
    #
    # @return [void]
    def train(inputs_list, targets_list)
      inputs  = inputs_list.expand_dims(0).transpose  # TODO
      targets = targets_list.expand_dims(0).transpose # TODO

      hidden_inputs  = @input_weights.dot(inputs)
      hidden_outputs = @activation_function.(hidden_inputs)

      final_inputs   = @output_weights.dot(hidden_outputs)
      final_outputs  = @activation_function.(final_inputs)

      output_errors  = targets - final_outputs
      hidden_errors  = @output_weights.transpose.dot(output_errors)

      @output_weights += @learning_rate * (output_errors * final_outputs * (1.0 - final_outputs)).dot(hidden_outputs.transpose)
      @input_weights  += @learning_rate * (hidden_errors * hidden_outputs * (1.0 - hidden_outputs)).dot(inputs.transpose)
    end

    ##
    # Queries the network.
    #
    # @return [Numo::DFloat]
    def query(inputs_list)
      inputs = inputs_list.expand_dims(0).transpose

      hidden_inputs  = @input_weights.dot(inputs)
      hidden_outputs = @activation_function.(hidden_inputs)

      final_inputs   = @output_weights.dot(hidden_outputs)
      final_outputs  = @activation_function.(final_inputs)
    end
  end # NeuralNetwork
end # MYONN


mnist_render.rb
#!/usr/bin/env ruby -Ilib
# This is free and unencumbered software released into the public domain.

require 'mnist'
require 'ruby2d' # See: https://www.ruby2d.com

# Constants
PIXEL_SIZE = 24
FONT_SIZE  = 336

# Globals
$current_index  = 0
$current_record = nil

# Window
set title: "Digit Renderer"
set background: 'white'
set width: MNIST::COLS * PIXEL_SIZE, height: MNIST::ROWS * PIXEL_SIZE # 672x672

def load_record(record_index)
  MNIST::Parser.open('data/mnist_train_100.csv') do |dataset|
    dataset.each_record.take(record_index + 1).last
  end
end

def render_digit
  _, pixels = $current_record
  clear
  MNIST::ROWS.times.map.with_index do |y|
    MNIST::COLS.times.map.with_index do |x|
      pixel = pixels[y, x]
      next if pixel.zero?
      value = 1.0 - (pixel / 255.0)
      color = Color.new([value, value, value, 1.0])
      Square.new(x: x * PIXEL_SIZE, y: y * PIXEL_SIZE, size: PIXEL_SIZE, color: color)
    end
  end
end

def render_label
  label, _ = $current_record
  clear
  Text.new(label.to_s,
    x: (Window.width  - FONT_SIZE) / 2.0 + FONT_SIZE * 0.25,
    y: (Window.height - FONT_SIZE) / 2.0,
    size: FONT_SIZE,
    color: 'black',
  )
end

on :key_up do |event|
  case k = event.key.to_sym
    when :escape then close
    when :up, :down, :left, :right
      $current_index += %i(down right).include?(k) ? 1 : -1
      $current_index = [0, $current_index].max
      $current_record = load_record($current_index)
      render_digit
    when :space
      render_label
  end
end

$current_record = load_record($current_index)
render_digit

show


lib/mnist/model.rb
# This is free and unencumbered software released into the public domain.

require 'myonn'
require 'mnist/parser'

module MNIST
  class Model
    INPUT_NODES   = 784 # 28x28 pixels
    HIDDEN_NODES  = 200
    OUTPUT_NODES  = 10  # (0..9)

    def self.load(filepath)
      self.new(nn: MYONN::NeuralNetwork.load(filepath))
    end

    def initialize(nn: nil, learning_rate: nil)
      @nn = nn || MYONN::NeuralNetwork.new(
        input_nodes:   INPUT_NODES,
        hidden_nodes:  HIDDEN_NODES,
        output_nodes:  OUTPUT_NODES,
        learning_rate: (learning_rate || 0.1).to_f,
      )
    end

    def save(filepath)
      @nn.save(filepath)
    end

    def train!(filepath, epochs: 1, limit: nil)
      zeros = Numo::DFloat.zeros(@nn.output_nodes)

      MNIST::Parser.open(filepath) do |dataset|
        epochs.times do
          count = 0
          dataset.each_record do |label, pixels|
            count += 1
            break if limit && count >= limit
            inputs = self.normalize(pixels)
            targets = zeros + 0.01
            targets[label] = 0.99
            @nn.train(inputs, targets)
          end
        end
      end
    end

    def validate(filepath)
      scorecard = []

      MNIST::Parser.open(filepath) do |dataset|
        dataset.each_record do |label, pixels|
          output = self.recognize(pixels)
          is_correct = (output == label)
          scorecard.append(is_correct ? 1 : 0)
        end
      end

      scorecard.sum / scorecard.size.to_f
    end

    def recognize(pixels)
      self.query(pixels).max_index()
    end

    def query(pixels)
      inputs = self.normalize(pixels)
      outputs = @nn.query(inputs)
    end

    def normalize(pixels)
      (pixels.reshape(INPUT_NODES) / 255.0 * 0.99) + 0.01
    end
  end # Model
end # MNIST


lib/mnist/parser.rb
# This is free and unencumbered software released into the public domain.

require 'numo/narray'

module MNIST
  ##
  # See: https://pjreddie.com/projects/mnist-in-csv/
  class Parser
    def self.open(filepath, &block)
      File.open(filepath, 'r') do |file|
        block.call(self.new(file))
      end
    end

    def initialize(file)
      @file = file
    end

    def each_record(&block)
      return enum_for(:each_record) unless block_given?

      @file.rewind
      @file.each_line do |csv_line|
        record = csv_line.chomp.split(',').map(&:to_i)
        label  = record.shift
        pixels = Numo::DFloat.cast(record).reshape!(*MNIST::SHAPE)
        block.call(label, pixels)
      end
    end

    def self.parse_record(csv_line)
      record = csv_line.chomp.split(',').map(&:to_i)
      [record.shift, record]
    end

    def normalize(values)
      (Numo::DFloat[*values] / 255.0 * 0.99) + 0.01
    end
  end # Parser
end # MNIST


mnist_draw.rb
#!/usr/bin/env ruby -Ilib
# This is free and unencumbered software released into the public domain.

require 'myonn'
require 'mnist'
require 'ruby2d' # See: https://www.ruby2d.com

# Constants
MNIST_MODEL = MNIST::Model.load('data/mnist_nn.yaml')
PIXEL_SIZE  = 24
FONT_SIZE   = 336

# Globals
$drawing = false
$pixels  = Numo::DFloat.zeros(MNIST::SHAPE)

# Window
set title: "Digit Recognizer"
set background: 'white'
set width: MNIST::COLS * PIXEL_SIZE, height: MNIST::ROWS * PIXEL_SIZE # 672x672

def reset
  $drawing = false
  $pixels = Numo::DFloat.zeros(MNIST::SHAPE)
end

def render_digit(pixels)
  clear
  MNIST::ROWS.times.map.with_index do |y|
    MNIST::COLS.times.map.with_index do |x|
      pixel = pixels[y, x]
      next if pixel.zero?
      value = 1.0 - (pixel / 255.0)
      color = Color.new([value, value, value, 1.0])
      Square.new(x: x * PIXEL_SIZE, y: y * PIXEL_SIZE, size: PIXEL_SIZE, color: color)
    end
  end
end

def render_label(label)
  clear
  Text.new(label.to_s,
    x: (Window.width  - FONT_SIZE) / 2.0 + FONT_SIZE * 0.25,
    y: (Window.height - FONT_SIZE) / 2.0,
    size: FONT_SIZE,
    color: 'black',
  )
end

on :key_up do |event|
  case event.key.to_sym
    when :escape then close
    when :space
      if $pixels
        output = MNIST_MODEL.query($pixels).reshape!(10).to_a
        puts output.map.with_index { |score, index| "%d=%.1f%%" % [index, score * 100] }.join("  ")
        digit = MNIST_MODEL.recognize($pixels)
        render_label(digit)
        $pixels = nil
      else
        reset
        clear
      end
  end
end

on :mouse_down do |event|
  case event.button.to_sym
    when :left then $drawing = true
  end
end

on :mouse_up do |event|
  case event.button.to_sym
    when :left then $drawing = false
  end
end

on :mouse_move do |event|
  if $drawing && $pixels
    x, y = event.x / PIXEL_SIZE, event.y / PIXEL_SIZE
    $pixels[y, x] = 255
    render_digit($pixels)
  end
end

reset

show
